### ISSUE TITLE
Data Augmentation Interface Designing

### ISSUE THREAD

**Issue by shijianjian** [kornia/kornia#495]
Since there are some use cases for the reproducible augmentations #476, #488.

I am wondering about how would the APIs look like. I think an easy approach is to have a flag to control if same augmentation would apply element-wise or batch-wise.

For normal use cases:
```python
rot = RandomRotation(360)
# If the same augmentation on the whole batch is not needed
rot(BATCH_OF_IMAGES, apply_same=False)
```
Or the same augmentation is needed, like in #488 
```python
rot = RandomRotation(360)
# If the same augmentation on the whole batch is needed
rot(BATCH_OF_IMAGES, apply_same=True)
```
Semantic Segmentation
```python
rot = RandomRotation(360)
# If the same augmentation on the different batches is needed
rot([image, mask], apply_same=True)
```
Pseudo logic like below:
```python
if input is tensor and BCHW:
    if apply_same:
        return aug_fn(input, param_set_a)
    else:
        return. aug_fn(input)
elif input is list:
    if apply_same:
        return aug_fn(input[0], param_set_a), ..., aug_fn(input[n], param_set_a)
    else:
        return aug_fn(input[0], param_set_a), ..., aug_fn(input[n], param_set_an)
```
Is any better idea or more use cases?

**Comment by edgarriba** [comment]
@shijianjian a couple of comments:
1) for the input data shape we want to randomize I see the following combinations: **B**TCHW, B**T**CHW, BT**C**HW, **B**CHW, B**C**HW. which suggests where to apply the generated parameters over B, T or C - so why just not adding `dim` flag specifying to which axis we want to apply the random parameters ? This convention will also go with pytorch standard (or `axis` as in numpy :)

2) I discourage this `rot([image, mask], apply_same=True)`. My argument against it is that the design, works for this specific case, but might over-complicate maintenance in the long term the more and more use cases we want to handle, plus, in the end could lead to reduce flexibility to the user. Instead, I would stick to the receive one single input and give the user the freedom to apply such transformation in each input of their pipeline (assuming we solve determinism in a proper way) /cc @anguelos

@shijianjian I actually liked your proposal in slack and keep functionalities simple so that users can do very customs things:
```python
class Augmentation():
    def __init__(self, params: torch.Tensor):
        # we might want to optimize params
        k1 = kornia.augmentation.RandomAffine(params)
        k2 = kornia.augmenataion.ColorJitter()
        k3 = kornia.augmentation.Normalize()

    def forward(self, image, mask):
        image = k3(k2(k1(image)))
        mask = k3((k1(image)))
        return image, mask
```

**Comment by shijianjian** [comment]
I agree that adding a parameter like ```apply_same_transform_on_dim``` looks good.

For semantic segmentation, it is actually a special case of BTCHW, which has a fixed T=2. User can input in (B, 2, C, H, W) with ```apply_same_transform_on_dim=1```.

In that logic, the ```nn.Module``` implementation shall be a wrapper to handle the dims, then for each transformation function, we gonna have a ```apply_same_on_batch``` param.
```python
if dim is not None:
    x_t = torch.transpose(x, 0, dim)
    x_t_transformed = torch.stack([transform(_x_t, apply_same_on_batch=True) for _x_t in x_t])
    x_transformed = torch.transpose(x_t_transformed, dim, dim)
else:
    x_t = torch.reshape(x, (-1, *x.size()[-3:]))
    x_t_transformed = torch.stack([transform(_x_t, apply_same_on_batch=False) for _x_t in x_t])
    x_transformed = torch.reshape(*size())
```

I am not sure the use case of B**C**HW, etc. Basically, intensity-based transforms required to have 3 color channels, for cases like B**C**HW, I do not think those intensity-based transforms will work.

**Comment by edgarriba** [comment]
I don't really get why do we need the for loop in `x_t_transformed`. If think that depending on the desired `dim` value, the generated parameters could be replicated so that the operator can be all cases applied in batch. BTW, I'd suggest to have by default `dim=0` being the transform applied to all the element in the batch.
